File: cole/day05_best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
# beats 98.22%

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        bestDiff = 0
        bestStart = prices[0]
        for day in prices[1:]:
            if day - bestStart > bestDiff:
                bestDiff = day - bestStart
            if day < bestStart:
                bestStart = day
        return bestDiff

File: cole/day05_best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.cpp
// beats 98.25%

#define MIN(a, b) a < b ? a : b
#define MAX(a, b) a > b ? a : b

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        auto it = prices.begin();
        int min = *it, max = 0;
        while (++it != prices.end())
        {
            min = MIN(min, *it);
            max = MAX(max, *it - min);
        }
        return max;
    }
};